src/OpenTelemetry/Metrics/ThreadStaticStorage.cs

```diff
@@ -0,0 +1,154 @@
+// <copyright file="ThreadStaticStorage.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Runtime.CompilerServices;
+
+namespace OpenTelemetry.Metrics
+{
+    internal class ThreadStaticStorage
+    {
+        private const int MaxTagCacheSize = 3;
+
+        [ThreadStatic]
+        private static ThreadStaticStorage storage;
+
+        private readonly TagStorage[] tagStorage = new TagStorage[MaxTagCacheSize + 1];
+
+        // .NET Metrics API allows for byte, short, int, long, float, double, decimal
+        private readonly Dictionary<Type, IDataPoint> pointT = new Dictionary<Type, IDataPoint>();
```

Comment:
Could we use `ValueType` instead of `Type` if we only allow built-in value types?

Reply:
Can you clarify a bit more?

I am reading your comments as...

You want the Key portion of the Dictionary to be a ValueType rather than a reference?

Dictionary<Type, IDataPoint> pointT;
pointT.Add(**typeof(T)**, dp);

So, we would need to find an equivalent ValueType for the Type class we get from typeof(T). I suppose we could use the Type.Guid property, etc...

What benefit are you expecting to get from this conversion in our case?

Reply:
I was suggesting to use the [ValueType](https://docs.microsoft.com/en-us/dotnet/api/system.valuetype?view=net-5.0) class instead of Type but I realized that wouldn't work as the `ValueType` class would actually have the value and not the type information. I was hoping to enforce that only value types could be used as keys but `ValueType` class is not suitable for this.